rsc/client/valve/replay.go

```diff
@@ -0,0 +1,92 @@
+package valve
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"strings"
+)
+
+var (
+	ErrDownloadRelpay = fmt.Errorf("failed to download the dota2 replay")
+)
+
+type ReplayDownloader interface {
+	// Download
+	// downloading the replay file is downloaded directly to the
+	// Valve replay proxy using the replayURL.
+	//
+	// replayURL contains the full raw URI to download the replay to the Valve replay proxy,
+	// for example http://replay153.valve.net/570/7132230434_1635105612.dem.bz2
+	// Based on that value, we can download the replay DotA2 file directly using that URL.
+	//
+	// destination parameter specifies the file destination to save the downloaded replay file.
+	Download(ctx context.Context, replayURL string, destination string) error
+}
+
+type replay struct {
+	httpClient *http.Client
+}
+
+func NewReplay(httpClient *http.Client) ReplayDownloader {
+	return &replay{
+		httpClient: httpClient,
+	}
+}
+
+func (r *replay) Download(ctx context.Context, replayURL string, destination string) error {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, replayURL, nil)
+
+	if err != nil {
+		return err
+	}
+
+	res, err := r.httpClient.Do(req)
+
+	if err != nil {
+		return err
+	}
+
+	defer res.Body.Close()
```

Comment:
style difference, but I'd rather remove newline whitespace here, keeping initial function, error handling and deferred function close to each other.

Reply:
Yeah, before I started working with my current lead, I already had the same coding style as you. However, my lead has encouraged me to add new spaces for better readability, although I think this is a matter of personal preference XD.

Reply:
updated to allign our style code